packages/ai/core/util/chat-store.ts

```diff
@@ -0,0 +1,602 @@
+import {
+  generateId,
+  generateId as generateIdFunction,
+  parsePartialJson,
+} from '.';
+import type {
+  ReasoningUIPart,
+  TextUIPart,
+  ToolInvocation,
+  ToolInvocationUIPart,
+  UIMessage,
+} from '../types';
+
+interface ChatStoreSubscriber {
+  onChatMessagesChanged?: (id: string) => void;
+  onChatStatusChanged?: (id: string) => void;
+  onChatErrorChanged?: (id: string) => void;
+  // onChatAdded?
+  // onChatRemoved?
+}
+
+export enum ChatStoreEvent {
+  ChatMessagesChanged = 'chat-messages-changed',
+  ChatStatusChanged = 'chat-status-changed',
+  ChatErrorChanged = 'chat-error-changed',
+  // ChatAdded = 'chat-added',
+  // ChatRemoved = 'chat-removed',
+}
+
+const ChatStoreEventMap = {
+  [ChatStoreEvent.ChatMessagesChanged]: 'onChatMessagesChanged',
+  [ChatStoreEvent.ChatStatusChanged]: 'onChatStatusChanged',
+  [ChatStoreEvent.ChatErrorChanged]: 'onChatErrorChanged',
+} as const;
+
+export interface ChatStoreInitialization {
+  chats?: Record<string, Pick<ChatState, 'messages'>>;
+  onChatStoreChange?: ({
+    event,
+    chatId,
+    state,
+  }: {
+    event: ChatStoreEvent;
+    chatId: string;
+    state: ChatState;
+  }) => void;
+}
+
+export interface ChatState {
+  status: 'submitted' | 'streaming' | 'ready' | 'error';
+  messages: UIMessage[];
+  step?: number;
+  activeResponse?: {
+    message: UIMessage;
+    partialState: {
+      textPart?: TextUIPart;
+      reasoningPart?: ReasoningUIPart;
+      reasoningTextDetail?: { type: 'text'; text: string; signature?: string };
+      tools?: Record<
+        string,
+        { step: number; text: string; index: number; toolName: string }
+      >;
+    };
+  };
+  error?: Error;
+}
+
+export class ChatStore {
+  private chats: Map<string, ChatState>;
+  private subscribers: Set<ChatStoreSubscriber>;
+  private onChatStoreChange?: ChatStoreInitialization['onChatStoreChange'];
+
+  constructor({ chats = {}, onChatStoreChange }: ChatStoreInitialization = {}) {
+    this.chats = new Map(
+      Object.entries(chats).map(([id, state]) => [
+        id,
+        {
+          messages: [...state.messages],
+          status: 'ready',
+          activeResponse: undefined,
+          error: undefined,
+          step: 0,
```

Comment:
is `step: 0` correct when there are pre-existing msgs (in case we want to keep `step`)?